frontend/src/pages/AccountReceive.vue

```diff
@@ -178,40 +179,92 @@ function useScan() {
 
     // Fetch announcements
     const overrides = { startBlock: startBlockLocal.value, endBlock: endBlockLocal.value };
-    let allAnnouncements: AnnouncementDetail[] = [];
-    try {
-      if (!signer.value) throw new Error('signer is undefined');
-      if (!userWalletAddress.value) throw new Error('userWalletAddress is undefined');
-      // When private key is provided in advanced mode, we fetch all announcements
-      if (advancedMode.value && scanPrivateKey.value)
-        allAnnouncements = await umbra.value.fetchAllAnnouncements(overrides);
-      else
-        allAnnouncements = await umbra.value.fetchSomeAnnouncements(signer.value, userWalletAddress.value, overrides);
-    } catch (e) {
-      scanStatus.value = 'waiting'; // reset to the default state because we were unable to fetch announcements
-      throw e;
-    }
 
     // Scan for funds
-    scanStatus.value = 'scanning';
     const spendingPubKey = chooseKey(spendingKeyPair.value?.publicKeyHex);
     const viewingPrivKey = chooseKey(viewingKeyPair.value?.privateKeyHex);
 
-    scanPercentage.value = 0;
-    filterUserAnnouncements(
-      spendingPubKey,
-      viewingPrivKey,
-      allAnnouncements,
-      (percent) => {
-        scanPercentage.value = Math.floor(percent);
-      },
-      (filteredAnnouncements) => {
-        userAnnouncements.value = filteredAnnouncements.sort(function (a, b) {
-          return parseInt(a.timestamp) - parseInt(b.timestamp);
-        });
+    // Wrapper for `filterUserAnnouncements` so we can await for the first batch of web workers to finish before
+    // creating new workers.
```

Comment:
What do you mean by "the first batch of web workers"? I'm not too familiar with them so would appreciate more details in the comments for my own understanding / to simplify future updates here. Is the number of web workers in the first batch equal to the max number of web workers allowed, or is it 1, or do we limit it to same ratio of the available web workers so we have others available for future incoming announcements?

Reply:
Thanks for the context in https://github.com/ScopeLift/umbra-protocol/pull/558#discussion_r1267213396! Moving the discussion here to avoid de-railing the scope of that thread.

I want to make sure I fully understand the flow. Specifically, I think we have:
1. First fetch 10k announcements, this may takes multiple HTTP requests to the subgraph/log so this happens serially with no web workers
2. Pass those 10k logs to `n` web workers to start decrypting/processing the logs, where `n < maxWorkers`. Question: How is `n` chosen though?
3. While 2 is processing, repeat step 1 to fetch 10k more logs with on web workers.
4. Pass the next batch to `m` web workers to start decrypting/processing the logs, where `m <= maxWorkers - n`. Question: How is `m` chosen? And how do we know if we need to reserve more workers for another batch of 10k?
5. Repeat the above steps until we've fetched all logs and allocated all workers to decrypting
6. As a worker finishes processing it's batch of logs, immediately show them in the table. Repeat this step until all logs are processed 

So let me know if that flow/understanding is correct. There's also some questions I wrote up in the above flow.

Also, if that's correct, it feels the answer to my question is "yes, we have to choose between 'stable table ordering when events come in' and 'sort table by timestamp'", because we can't guarantee that e.g. step 2 finishes before step 4. Is that also correct?

Reply:
> 2. Pass those 10k logs to n web workers to start decrypting/processing the logs, where n < maxWorkers. Question: How is n chosen?

I believe `n` is maxWorkers allowed. All of these are done inside `frontend/src/worker/worker.ts` file, I believe added with this PR https://github.com/ScopeLift/umbra-protocol/pull/289

> 3. While 2 is processing, repeat step 1 to fetch 10k more logs with no web workers.

No, we actually go on to fetch rest of the announcements with no web workers(except the ones that were created in step 2), doesn't matter if there's 1000 more to fetch or 30000 more.

4. First batch of web workers has to finish stop scanning before we do anything else because creating additional workers actually will slow down the scan of first 10,000 announcements. But meanwhile we can keep fetching until we've fetch all of the rest of the announcements.

5. Web workers finish scanning 10,000 announcements.

6. New `n` webworkers are created to process the rest of the announcements (maybe 1000, maybe 30000+, doesn't matter, this is the final scan we will do).

Reply:
We only refresh the receive table when the first scan finishes and when the second/last scan finishes. And they are ordered by timestamp inside `filterUserAnnouncements`

Reply:
Ok getting closer! Updated understanding:

1. First fetch 10k announcements, this may takes multiple HTTP requests to the subgraph/node so this happens serially with no web workers.
2. Pass those 10k logs to `maxWorkers` web workers to start decrypting/processing the logs.
3. While 2 is processing, repeat step 1 to fetch *all* remaining announcement logs.
4. Wait:
    1. For step (2) to complete, and show any results in the table. At this point, all web workers have no jobs.
    2. For step (3) to complete. At this point, the app has all announcement logs.
5. Give all remaining logs to `maxWorkers` to process. Wait for them and show the rest in the table. 

Questions:
1. Are the web workers from (2) shut down then re-spawned in (5), or do they stay active but idle until step (5)?
2. Does step 5 update the table as workers finish processing, or does it wait for *all* remaining logs to be processed then update the table?

Reply:
That's correct!

> 1. Are the web workers from (2) shut down then re-spawned in (5), or do they stay active but idle until step (5)?

I haven't changed `worker.ts` behavior, so they shut down and then re-spawn in (5).

> 2. Does step 5 update the table as workers finish processing, or does it wait for all remaining logs to be processed then update the table?

Table is updated only when all remaining logs are processed and done.

Reply:
Lmk if you know an easy or obvious way of doing this
> update the table contents with additional records after they're found rather than completely reseting/re-rendering the table.

A short search didn't reveal anything straightforward

Reply:
Now using [watchEffect](https://vuejs.org/guide/essentials/watchers.html#watcheffect) to accomplish this

Reply:
> That's correct!

Great, let's make sure this flow is documented somewhere, a big comment block in whatever file seems best is ok (maybe `worker.ts`?)

Reply:
> A short search didn't reveal anything straightforward

Yea I don't know of anything offhand, a watcher seems ok if that works. 

Thanks for all the back and forth here!

Reply:
Thank you @alexkeating !